library/core/src/fmt/mod.rs

```diff
@@ -231,6 +244,243 @@ impl<W: Write + ?Sized> Write for &mut W {
     }
 }
 
+/// The signedness of a [`Formatter`] (or of a [`FormattingOptions`]).
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+#[unstable(feature = "formatting_options", issue = "118117")]
+pub enum Sign {
+    /// Represents the `+` flag.
+    Plus,
+    /// Represents the `-` flag.
+    Minus,
+}
+
+/// Specifies whether the [`Debug`] trait should use lower-/upper-case
+/// hexadecimal or normal integers.
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+#[unstable(feature = "formatting_options", issue = "118117")]
+pub enum DebugAsHex {
+    /// Use lower-case hexadecimal integers for the `Debug` trait (like [the `x?` type](../../std/fmt/index.html#formatting-traits)).
+    Lower,
+    /// Use upper-case hexadecimal integers for the `Debug` trait (like [the `x?` type](../../std/fmt/index.html#formatting-traits)).
+    Upper,
+}
+
+/// Options for formatting.
+///
+/// `FormattingOptions` is a [`Formatter`] without an attached [`Write`] trait.
+/// It is mainly used to construct `Formatter` instances.
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+#[unstable(feature = "formatting_options", issue = "118117")]
+pub struct FormattingOptions {
+    sign: Option<Sign>,
+    sign_aware_zero_pad: bool,
+    alternate: bool,
+    fill: char,
+    align: Option<Alignment>,
+    width: Option<usize>,
+    precision: Option<usize>,
+    debug_as_hex: Option<DebugAsHex>,
+}
+
+impl FormattingOptions {
+    /// Construct a new `FormatterBuilder` with the supplied `Write` trait
+    /// object for output that is equivalent to the `{}` formatting
+    /// specifier:
+    ///
+    /// - no flags,
+    /// - filled with spaces,
+    /// - no alignment,
+    /// - no width,
+    /// - no precision, and
+    /// - no [`DebugAsHex`] output mode.
+    #[unstable(feature = "formatting_options", issue = "118117")]
+    pub fn new() -> Self {
```

Comment:
It might be useful to make these functions `const` where possible.

Reply:
I agree, I made all methods/functions const. Some of them receive `this` or some other parameter as `&mut`, which [is still unstable](https://github.com/rust-lang/rust/issues/57349) – I'm not sure whether this is a problem here and we shouldn't make those methods const, as there are still some bugs mentioned in the tracking issue.

Reply:
Maybe it's best to not make those &mut methods const for now, but leave it as an unresolved question in the tracking issue so we don't forget.

Reply:
I'm not too sure how helpful the remaining `const` methods are (all the setters receive `self` as `&mut`, so I think the only thing that can be done in a const context is to construct default formatting options via `FormattingOptions::new`, which doesn't sound very exciting), but that seems like an acceptable compromise for now.

I've removed `const`ness from the affected methods and added this as an unresolved question to the tracking issue.